lewis/adapters/modbus.py

```diff
@@ -0,0 +1,450 @@
+        :return: ModbusTCPFrame instance that represents an exception
+        """
+        frame = deepcopy(self)
+        frame.length = 3
+        frame.fcode += 0x80
+        frame.data = bytearray(chr(code))
+        return frame
+
+    def create_response(self, data=None):
+        """
+        Create a response frame based on this frame.
+
+        :param data: Data section of response as bytearray. If None, request data section is kept.
+        :return: ModbusTCPFrame instance that represents a response
+        """
+        frame = deepcopy(self)
+        if data is not None:
+            frame.data = data
+        frame.length = 2 + len(frame.data)
+        return frame
+
+
+class ModbusProtocol(object):
+    """
+    StateMachine based Modbus Protocol parser.
+    """
+
+    def _init_state(self):
+        self._request = None
+        self._response = None
+        self._exception = None
+
+    def __init__(self, handler, databank):
+        self._init_state()
+
+        self.handler = handler
+        self.databank = databank
+        self.buffer = bytearray()
+
+        self.protocol = StateMachine({
+            'initial': 'await_request',
+            'transitions': OrderedDict([
+                (('await_request', 'check_fcode'), lambda: self._request is not None),
+                (('check_fcode', 'send_response'), lambda: self._exception is not None),
+
+                (('check_fcode', 'read_bit'),
+                 lambda: self._request.fcode in
+                    (MBFC.READ_COILS, MBFC.READ_DISCRETE_INPUTS)),
+                (('check_fcode', 'read_word'),
+                 lambda: self._request.fcode in
+                    (MBFC.READ_HOLDING_REGISTERS, MBFC.READ_INPUT_REGISTERS)),
+                (('check_fcode', 'write_coil'),
+                 lambda: self._request.fcode == MBFC.WRITE_SINGLE_COIL),
+                (('check_fcode', 'write_register'),
+                 lambda: self._request.fcode == MBFC.WRITE_SINGLE_REGISTER),
+
+                (('read_bit', 'send_response'), lambda: True),
+                (('read_word', 'send_response'), lambda: True),
+                (('write_coil', 'send_response'), lambda: True),
+                (('write_register', 'send_response'), lambda: True),
+                (('send_response', 'await_request'), lambda: True),
+
+                # Print unknown packets
+                (('check_fcode', 'print_packet'), lambda: True),
+                (('print_packet', 'await_request'), lambda: True),
+            ])
+        })
+        self.protocol.bind_handlers_by_name(self, prefix=['_on_', '_in_', '_after_'])
+        self.protocol.process()  # Cycle into initial state
+
+    def parse(self, data=None):
+        """
+        Parse as much of current buffer as possible.
+
+        :param data: Optionally append given data to buffer prior to parsing
+        """
+        self.buffer.extend(bytearray(data or ''))
+
+        # Process until two or three FSM cycles stay in the same state
```

Comment:
Ah, I think I misunderstood the question a bit (thought it was about the buffer)... but still sort of answered it, partially, in a roundabout way.

Now that `ModbusTCPFrame` always returns a complete request, the condition this is really waiting for is to spend more than one cycle in `await_request`. Because this means the buffer no longer contains any complete requests.

This is necessary because sometimes more than one request comes in as one big chunk of requests. If you try using the asynchronous client example, instead of the synchronous one, you'll see that ALL the requests come in as one big packet in that one. I haven't seen partial requests yet, but in principle this ModbusAdapter would support that.

I wouldn't really want a `parse_complete` state since... well, what do you do in that? If you're done with one request, you should wait for the next one. Or go ahead and process the next one, if you already have enough data for one.